### ISSUE TITLE
docker login creates keychain entries with '///' on OSX

### ISSUE THREAD

**Issue by serenitus** [moby/moby#33043]
**Description**

I'm using AWS ECR (if it matters). When performing a `docker login` command the credentials get written to the OSX keychain. However, they're getting written with a triple-slash in the URL and this then breaks builds that pull from the protected registry (obviously having a locally cached image masks this error).

![image](https://cloud.githubusercontent.com/assets/3412351/25745317/45813e54-3196-11e7-9075-e1e316dc7a51.png)

**Steps to reproduce the issue:**

Don't think the the AWS ECR part is relevant but including it anyway.

1. Run: `aws ecr get-login --profile prod --region us-east-1` - this will produce a docker login command
2. Run the command obtained from step 1.
3. Open OSX keychain and search for 'Docker' 

**Describe the results you received:**

I see an incorrect URL in the keychain as per the attached screenshot.

**Describe the results you expected:**

A URL with only two slashes.

**Output of `docker version`:**

```
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Tue Mar 28 00:40:02 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Fri Mar 24 00:00:50 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 13
Server Version: 17.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.13-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952 GiB
Name: moby
ID: 3TM5:M3MZ:KIVX:WF65:CLIT:FQZW:K622:UWPI:R5TX:K7WU:RZGJ:AETB
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 16
 Goroutines: 27
 System Time: 2017-05-05T12:28:47.385077075Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

macOS Sierra 10.12.4

**Comment by cpuguy83** [comment]
Is step 1 producing a `docker login` command?
What exactly does it produce (redacted as neccessary of course).

**Comment by serenitus** [comment]
```
docker login -u AWS -p <redacted> -e none https://143850348497.dkr.ecr.us-east-1.amazonaws.com
```

**Comment by thaJeztah** [comment]
Ok the problem looks to be that when _storing_ the credentials, the scheme is omitted. As a result, keychain stores the credentials with an "empty" hostname, and takes the hostname that _is_ provided as _path_. Basically, the keychain stores;

```
http://<empty>/143850348497.dkr.ecr.us-east-1.amazonaws.com
```

You can emulate this, by manually storing a credential;

```bash
echo '{"ServerURL": "https://example.com/", "Username":"foobar", "Secret":"password"}' | docker-credential-osxkeychain store
```

This correctly stores `https://example.com`

Wheras;

```bash
echo '{"ServerURL":"example.com","Username":"foobar","Secret":"password"}' | docker-credential-osxkeychain store
```

Stores `http:///example.com`

The part where this goes "wrong" is; [`registry.ConfigureAuth()`](https://github.com/docker/cli/blob/862649707ef2b901fc0535895de7456fc4c9c0c3/cli/command/registry.go#L84-L86], which uses [`ConvertToHostname()`](https://github.com/docker/cli/blob/master/vendor/github.com/docker/docker/registry/auth.go#L210-L223) to "normalise" the ServerURL, and strip the scheme.

Interestingly, when using the _default_ registry, the scheme is not stripped, and a full URL (`https://index.docker.io/v1/`) is stored
https://github.com/docker/cli/blob/4f9ac4899f9329e5d942963cdfb4af4a9c360fef/vendor/github.com/docker/docker/registry/config.go#L45-L46

I think that, if it's a valid URL, we should keep the scheme. If no scheme is provided, possibly default to `https://` (not sure if that would cause issues with insecure registries, but IIRC, it would automatically fall back to `http://`)

**Comment by thaJeztah** [comment]
ping @simonferquel @n4ss @aaronlehmann

**Comment by aaronlehmann** [comment]
> I think that, if it's a valid URL, we should keep the scheme. If no scheme is provided, possibly default to https:// (not sure if that would cause issues with insecure registries, but IIRC, it would automatically fall back to http://)

Not sure if it's different with credential helpers, but when saving to a file, we only store the scheme for the default registry (which is an unfortunate exception). I don't know if we can change this without breaking backward compatibility.

**Comment by thaJeztah** [comment]
Yes, I was wondering the same. Unfortunately there's a bit too much mangling of these values going on, and need to look into that if we can

- reduce the amount of modifications of these values
- do so in a backward-compatible way (which _may_ involve having to retry with "old" values)

**Comment by thaJeztah** [comment]
Worst case probably is that people have to re-authenticate (thinking out loud)

**Comment by n4ss** [comment]
Not sure how we can avoid breaking compatibility but that'd be great to default to https directly here before calling the cred helpers.
